runners/core-java/src/main/java/org/apache/beam/runners/core/TimerInternals.java

```diff
@@ -171,33 +173,63 @@
 
     public abstract StateNamespace getNamespace();
 
+    /** Timestamp, at which the timer fires. */
     public abstract Instant getTimestamp();
 
+    /**
+     * Timestamp the timer assigns to outputted elements from
+     * {@link org.apache.beam.sdk.transforms.DoFn.OnTimer} method. For event time timers, output
+     * watermark is held at this timestamp until the timer fires.
+     * */
+    public abstract Instant getOutputTimestamp();
+
     public abstract TimeDomain getDomain();
 
     // When adding a new field, make sure to add it to the compareTo() method.
 
     /**
-     * Construct a {@link TimerData} for the given parameters, where the timer ID is automatically
-     * generated.
+     * Construct a {@link TimerData} for the given parameters.
+     */
+    public static TimerData of(
```

Comment:
Now that we have two `Instant` value parameters, it is a good time to switch this to use `@AutoValue.Builder`.